File: WEEK_1/supplementary_code/W1Q1_vary_h0.py
import matplotlib.pyplot as plt   # import package for plot

# Physical parameters
Q_in = 1     #inflow rate
A_t = 1      # tank cross-sectional area (m^2)
A_out = 0.1  # outlet area (m^2)
g = 9.81     # gravity

# Simulation time parameters
T = 100       # total simulation time 
N = 1000      # number of time steps
dt = T / N    # time step

# outflow rate Q_out
def Q_out(h):
    return A_out * (2 * g * h) ** 0.5

plt.figure(figsize=(10, 5))

# loopover different values of intial water level
for h0 in [0, 2, 5, 8, 10]:

    # lists to store the solution 
    times = (N+1)*[0]
    h = (N+1)*[0]
    h[0] = h0
    
    for i in range(N):
        dh = dt * (Q_in - Q_out(h[i]))
        h[i+1] = h[i] + dh 
        times[i+1] = (i+1)*dt
    
    plt.plot(times, h, label = 'h0 = '+str(h0))

plt.xlabel("Time (s)")
plt.ylabel("Water Level (m)")
plt.title("Water Level Over Time for Different h0")
plt.legend()
plt.grid()
plt.show()
    

File: WEEK_1/supplementary_code/W1Q1_vary_h0.m
% Function to define the inflow rate function
Q_in = @(t) 1;
% Function to calculate the outflow rate based on the water level and outlet area
Q_out = @(A_out, h) sqrt(2*9.81 * max(h,0)) * A_out;
A_out = 0.1; % Outlet area (m2)
A_t = 1; % Tank cross-sectional area (m2)

% Simulation time parameters
T = 100; % Total simulation time (seconds)
N = 1000; % Number of time steps
% Calculate the time step size
dt = T / N;

% Create figure for plotting
figure()
grid on
hold on

% Loop over different initial water levels
for h0 = [0, 1, 2, 5, 8]
  % Create arrays to store data for plotting
  times = zeros(1, N + 1);
  h = zeros(1, N + 1);
  h(1) = h0;
  t = 0;
  % Simulate the water level over time
  for i = 1:N
      t = t + dt;
      h(i+1) = h(i) + (Q_in(t) - Q_out(A_out, h(i))) * dt;
      times(i+1) = t;
  end
  plot(times, h, '-')
end

legend('h0=0', 'h0=1', 'h0=2', 'h0=5', 'h0=8')

